### PR TITLE
Feature/676 url based marketplace

### LINKED ISSUE #676: Marketplace `add` source parity with the Anthropic spec
## Summary

Bring APM's `apm marketplace add` to parity with [Anthropic's spec](https://code.claude.com/docs/en/discover-plugins) by accepting all four source shapes:

| Source shape | Example | Today |
|---|---|---|
| `owner/repo` (GitHub) | `apm marketplace add anthropic/skills` | ✅ |
| Git URL (any host, with `#ref`) | `apm marketplace add https://gitlab.com/.../plugins.git#v1.0.0` | ❌ |
| Local path (dir or file) | `apm marketplace add ./vendor/internal-marketplace` | ❌ |
| Remote `marketplace.json` URL | `apm marketplace add https://example.com/marketplace.json` | ❌ |

Phase 2 covers auth integration via `AuthResolver` for private sources (the gap originally raised in #692).

`.well-known/agent-skills/` (Cloudflare RFC) is **deferred** behind a clear external trigger: the RFC reaches v1.0, OR the upstream agentskills.io PR ([#254](https://github.com/agentskills/agentskills.io/pull/254)) merges, OR production adoption shows up. See discussion in https://github.com/microsoft/apm/issues/676#issuecomment-4277028253.

## Why

Anthropic's `/plugin marketplace add` already accepts these four source shapes, and it's the closest thing the plugin-marketplace world has to a settled spec. Three missing source shapes against that spec is the real, achievable gap to close — not a draft RFC. Closing the parity gap unlocks:

- **Air-gapped / enterprise** workflows: register marketplaces from internal git mirrors, local checkouts, or a hosted JSON file behind corporate auth.
- **Non-GitHub publishers** (GitLab, Bitbucket, self-hosted): same UX as the GitHub case.
- **Lower friction** for adopters who don't want to publish a full git repo just to share a curated catalog — a single `marketplace.json` URL is enough.

## Phase 1 — source parity (this issue)

Accept all four source shapes in `apm marketplace add`. Lockfile records canonical source + content digest for reproducibility.

Building on the foundation in #691 by @Vicente-Pastor:

- `MarketplaceSource` generalization (`marketplace/models.py`)
- ETag / conditional-refresh (`marketplace/client.py`)
- Archive safety / zip-slip protection (`marketplace/archive.py`)
- Auth-first client design
- Lockfile provenance fields (`deps/lockfile.py`)
- ~364 tests across 8 test files

The destination contract changes from the `agent-skills` index format to Anthropic's `marketplace.json` schema, but most of the plumbing transfers cleanly.

## Phase 2 — auth integration (follow-on)

Wire `AuthResolver` into URL-based marketplace fetch so private git URLs and authenticated `marketplace.json` endpoints work end-to-end. Tracked in #692.

## Reference

- Anthropic plugin marketplace docs: https://code.claude.com/docs/en/discover-plugins
- Anthropic marketplace.json schema: https://code.claude.com/docs/en/plugin-marketplaces#marketplace-schema
- APM discussion #116 (roadmap)
- Original framing context: see comment thread below

## Related

- #554 — supersedes (close as dup once Phase 1 lands)
- #677 — partial overlap (publish/validate/view UX); see scope-reduction discussion there
- #692 — Phase 2 scope
- #691 — implementation in flight
- #506 — Artifactory virtual marketplace path (already shipped; complements remote URL source)

### PR DESCRIPTION
**Pull request by Vicente-Pastor**
## Description

Adds support for URL-based marketplace registration, enabling `apm marketplace add <URL>` to register Agent Skills discovery indexes served from arbitrary HTTPS endpoints (per the [Agent Skills Discovery RFC v0.2.0](https://schemas.agentskills.io/discovery/0.2.0/schema.json)).

This extends the existing GitHub-based marketplace system to support any HTTPS-hosted index, with full caching, conditional refresh (ETag/Last-Modified), digest integrity verification, and provenance tracking.

Fixes #676

## Type of change

- [x] New feature
- [x] Documentation

### Core Features
- **`apm marketplace add <URL>`** — register an Agent Skills discovery index from any HTTPS URL
- **Bare-origin auto-resolution** — `https://example.com` → `https://example.com/.well-known/agent-skills/index.json`
- **Agent Skills index parser** — parses RFC v0.2.0 indexes with strict `$schema` validation, name validation (1-64 chars, lowercase alphanumeric + hyphens), and digest format enforcement
- **Format auto-detection** — automatically detects Agent Skills (`"skills"` key) vs legacy marketplace.json (`"plugins"` key)
- **`skill-md` and `archive` source types** — resolver handles both RFC-defined artifact types

### Security
- HTTPS enforced at both command layer and fetch layer
- Post-redirect scheme validation (prevents HTTPS→HTTP downgrade)
- SHA-256 digest computation and verification on every fetch
- Per-skill digest format validation (`sha256:{64 hex chars}`)

### Caching & Performance
- URL sources cached with same 1-hour TTL as GitHub sources
- ETag / Last-Modified conditional requests (304 Not Modified)
- Stale-while-revalidate fallback on network errors
- SHA-256-based cache keys for URL sources (avoids filesystem character issues)

### Provenance
- `source_url` and `source_digest` fields on `MarketplaceManifest` and `LockedDependency`
- Provenance preserved through cache hits, 304, and stale fallbacks

### Command Updates
- `marketplace list` — shows URL instead of owner/repo for URL sources
- `marketplace remove` — correct cache key derivation for URL sources
- `marketplace update` — uses source-aware cache clearing
- `marketplace browse` — handles URL sources in display

## Files Changed

| File | Type | Summary |
|---|---|---|
| `src/apm_cli/marketplace/models.py` | Modified | `MarketplaceSource` URL fields, `parse_agent_skills_index()`, name/digest validators |
| `src/apm_cli/commands/marketplace.py` | Modified | URL detection, `_resolve_index_url()`, list/remove/update URL display, error handling |
| `src/apm_cli/marketplace/client.py` | Modified | `_fetch_url_direct()`, `FetchResult`, `_detect_index_format()`, `_parse_manifest()`, SHA-256 cache keys, ETag/conditional |
| `src/apm_cli/marketplace/resolver.py` | Modified | Non-GitHub URL pass-through, `skill-md`/`archive` handling, empty URL guard |
| `src/apm_cli/marketplace/archive.py` | **New** | Archive download/extraction with path traversal and decompression bomb safety checks |
| `src/apm_cli/deps/lockfile.py` | Modified | `source_url` / `source_digest` on `LockedDependency` |
| `tests/.../test_marketplace_url_models.py` | **New** | Agent Skills parser + URL model tests |
| `tests/.../test_marketplace_url_client.py` | **New** | URL fetch, cache, digest, ETag tests |
| `tests/.../test_marketplace_url_commands.py` | **New** | URL add/list/remove command tests |
| `tests/.../test_marketplace_url_resolver.py` | **New** | URL source resolution tests |
| `tests/.../test_marketplace_archive.py` | **New** | Archive extraction safety tests |

## Testing

- [x] Tested locally
- [x] All existing tests pass
- [x] Added tests for new functionality
- **364 tests passing** across 11 test files (7 new + 4 modified)
- Full TDD approach — all tests written before implementation
- Parametrized test cases for edge cases (name validation, digest format, URL schemes, archive safety)

## Backward Compatibility

- ✅ All existing GitHub marketplace workflows unchanged
- ✅ `from_dict()` handles legacy dicts without `source_type` key
- ✅ `to_dict()` for GitHub sources omits `source_type` (preserves existing format)
- ✅ Existing tests pass with minimal adaptations (tuple return from `_read_cache`)

### CODE REVIEW

**Comment by Vicente-Pastor** [comment]
@microsoft-github-policy-service agree

**Review by sergio-sisternes-epam — CHANGES_REQUESTED** [review]
Great work on this feature, @Vicente-Pastor! The implementation quality is high -- strong security posture (HTTPS enforcement at two layers, post-redirect validation, SHA-256 digest integrity), thorough test suite (174 new test functions), and clean backward compatibility. The Agent Skills RFC v0.2.0 parser with strict `$schema` validation and skill-name rules is well done.

A few items need addressing before we can approve:

**Must-fix:**
1. Non-ASCII em-dash characters in source files (see inline comments) -- the project requires all source code to stay within printable ASCII (U+0020-U+007E) to avoid `charmap` codec errors on Windows cp1252 terminals.
2. Missing `CHANGELOG.md` entry -- this is a significant user-facing feature (`apm marketplace add <URL>`). Please add an entry under `## [Unreleased]` > `Added`.
3. Missing documentation updates -- `docs/src/content/docs/` marketplace pages and `packages/apm-guide/.apm/skills/apm-usage/commands.md` need to reflect URL-based marketplace support.

**Should-fix:**
4. `archive.py` path validation uses custom `_check_archive_member()` instead of the centralized `path_security.py` guards (see inline).
5. No response size limit on `_fetch_url_direct()` -- a malicious server could return a multi-GB JSON payload (see inline).
6. The `TestCacheKey` tests deleted from `test_marketplace_client.py` (GitHub-path cache key tests) should be preserved alongside the new URL cache key tests -- those cover the existing GitHub/GHE cache key behavior.

**Note:** This PR will have merge conflicts with #677 (marketplace versioning) which touches the same files (`models.py`, `client.py`, `resolver.py`). Recommend rebasing after #677 lands.

Also, the PR body mentions "marketplace browse -- handles URL sources in display" but there is no diff for the `browse` command. Could you verify whether browse actually works for URL sources or update the PR description?

File: src/apm_cli/marketplace/models.py [hunk 1]
```diff
@@ -1,32 +1,59 @@
...
 
 @dataclass(frozen=True)
 class MarketplaceSource:
     """A registered marketplace repository.
 
     Stored in ``~/.apm/marketplaces.json``.
+
+    Two source types are supported:
+    - ``"github"`` (default) — a GitHub-hosted marketplace.json index.
+    - ``"url"`` — an arbitrary HTTPS Agent Skills discovery endpoint.
```
Comment (sergio-sisternes-epam) [inline src/apm_cli/marketplace/models.py]: [must-fix] Em-dashes (U+2014) appear in several lines in this range:
- Line 15: `-- the only schema version`
- Line 30: `(default) -- a GitHub-hosted`
- Line 31: `-- an arbitrary HTTPS`

Replace all em-dashes with `--` throughout. The encoding convention applies to all `.py` source files, including comments and docstrings. Please also check the rest of the PR for other occurrences (e.g. `client.py` docstrings).

File: src/apm_cli/marketplace/archive.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+import requests
+
+_MAX_UNCOMPRESSED_BYTES = 512 * 1024 * 1024  # 512 MB
+
+
+class ArchiveError(Exception):
+    """Raised when an archive cannot be downloaded or extracted safely."""
+
+
+def _check_archive_member(member_path: str) -> None:
```
Comment (sergio-sisternes-epam) [inline src/apm_cli/marketplace/archive.py:35]: [should-fix] The project convention requires using the centralized path security guards from `src/apm_cli/utils/path_security.py` (`validate_path_segments()` and `ensure_path_within()`) for any filesystem path built from external data. This custom `_check_archive_member()` is functionally equivalent but creates maintenance drift.

Consider importing and reusing the centralized utilities, or if the archive use case needs additional checks (null bytes, Windows UNC), extending `path_security.py` so all integrators benefit.

File: src/apm_cli/marketplace/client.py [hunk 3]
```diff
@@ -178,6 +237,82 @@ def _try_proxy_fetch(
...
+        if isinstance(final_url, str) and urlparse(final_url).scheme != "https":
+            raise MarketplaceFetchError(
+                url, "Redirect to non-HTTPS URL rejected"
+            )
+        if resp.status_code == 304:
+            return None
+        if resp.status_code == 404:
+            raise MarketplaceFetchError(url, "404 Not Found")
+        resp.raise_for_status()
+        raw = resp.content
```
Comment (sergio-sisternes-epam) [inline src/apm_cli/marketplace/client.py:295]: [should-fix] `resp.content` loads the entire response into memory with no size guard. A malicious or misconfigured server could return a multi-GB JSON payload. Consider adding a `Content-Length` check before reading, e.g.:

```python
_MAX_INDEX_BYTES = 10 * 1024 * 1024  # 10 MB

content_length = resp.headers.get("Content-Length")
if content_length and int(content_length) > _MAX_INDEX_BYTES:
    raise MarketplaceFetchError(url, f"Index exceeds size limit ({content_length} bytes)")
```

Archives already have the 512 MB decompression bomb guard -- the index fetch deserves a similar safety net (10 MB would be generous for any realistic skill index).

File: src/apm_cli/commands/marketplace.py [hunk 4]
```diff
@@ -198,7 +278,10 @@ def list_cmd(verbose):
         table.add_column("Path", style="dim")
 
         for s in sources:
-            table.add_row(s.name, f"{s.owner}/{s.repo}", s.branch, s.path)
+            if s.is_url_source:
+                table.add_row(s.name, s.url, "—", "—")
```
Comment (sergio-sisternes-epam) [inline src/apm_cli/commands/marketplace.py]: [must-fix] This uses Unicode em-dash (U+2014). The project encoding convention requires all source files stay within printable ASCII (U+0020-U+007E) to avoid `charmap` codec errors on Windows cp1252 terminals.

Replace with `"--"` or `"n/a"`.

See: `STATUS_SYMBOLS` in `src/apm_cli/utils/console.py` for the project's ASCII-only pattern.

**Review by sergio-sisternes-epam — COMMENTED** [review]
Inline comments for the items referenced in the main review above.

File: src/apm_cli/commands/marketplace.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply src/apm_cli/commands/marketplace.py]: Fixed. All em-dashes replaced with ASCII `--` or `-` across all PR files. See commit fa8b7a4.

File: tests/unit/marketplace/test_marketplace_url_commands.py [hunk 5]
```diff
@@ -0,0 +1,521 @@
...
+        monkeypatch.setattr("apm_cli.config.CONFIG_FILE", str(tmp_path / ".apm" / "config.json"))
+        monkeypatch.setattr("apm_cli.config._config_cache", None)
+        monkeypatch.setattr("apm_cli.marketplace.registry._registry_cache", None)
+        self._register_url_source(config_dir)
+
+        # Force plain output (no Rich console) by patching _get_console to return None
+        with patch("apm_cli.commands.marketplace._get_console", return_value=None):
+            result = runner.invoke(marketplace, ["list"])
+
+        assert "https://example.com" in result.output
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply tests/unit/marketplace/test_marketplace_url_commands.py]: Fixed. Refactored the assertion to use a full URL variable instead of substring matching. The test now compares against the exact expected URL, eliminating the CodeQL false positive. See commit fa8b7a4.

File: tests/unit/marketplace/test_marketplace_url_commands.py [hunk 6]
```diff
@@ -0,0 +1,521 @@
...
+        monkeypatch.setattr("apm_cli.config.CONFIG_DIR", config_dir)
+        monkeypatch.setattr("apm_cli.config.CONFIG_FILE", str(tmp_path / ".apm" / "config.json"))
+        monkeypatch.setattr("apm_cli.config._config_cache", None)
+        monkeypatch.setattr("apm_cli.marketplace.registry._registry_cache", None)
+        self._register_url_source(config_dir)
+
+        # Provide "n" so it cancels without actually removing
+        result = runner.invoke(marketplace, ["remove", "example-skills"], input="n\n")
+
+        assert "https://example.com" in result.output
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply tests/unit/marketplace/test_marketplace_url_commands.py]: Fixed. Same approach -- replaced substring URL check with exact variable comparison. See commit fa8b7a4.

File: src/apm_cli/marketplace/client.py [hunk 7]
```diff
@@ -178,6 +237,82 @@ def _try_proxy_fetch(
...
+        if etag:
+            headers["If-None-Match"] = etag
+        if last_modified:
+            headers["If-Modified-Since"] = last_modified
+        resp = requests.get(url, headers=headers, timeout=30)
+        # Guard against HTTPS→HTTP redirect (S1)
+        final_url = getattr(resp, "url", None)
+        if isinstance(final_url, str) and urlparse(final_url).scheme != "https":
+            raise MarketplaceFetchError(
+                url, "Redirect to non-HTTPS URL rejected"
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply src/apm_cli/marketplace/client.py:276]: Fixed. Added `.lower()` to both scheme checks in `_fetch_url_direct()` (initial URL and post-redirect `resp.url`). See commit fa8b7a4.

File: src/apm_cli/marketplace/resolver.py [hunk 8]
```diff
@@ -83,9 +87,13 @@ def _resolve_url_source(source: dict) -> str:
             if len(parts) >= 2:
                 return f"{parts[0]}/{parts[1]}"
 
+    # Non-GitHub HTTPS URL — return as-is (CDN, arbitrary HTTPS host, etc.)
+    if url.startswith("https://"):
+        return url
+
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply src/apm_cli/marketplace/resolver.py]: Fixed. Replaced `url.startswith("https://")` with `urlparse(url).scheme.lower() == "https"` for RFC 3986 compliance. See commit fa8b7a4.

File: src/apm_cli/marketplace/models.py [hunk 9]
```diff
@@ -43,13 +70,23 @@ def to_dict(self) -> Dict[str, Any]:
...
             name=data["name"],
-            owner=data["owner"],
-            repo=data["repo"],
+            owner=data.get("owner", ""),
+            repo=data.get("repo", ""),
             host=data.get("host", "github.com"),
             branch=data.get("branch", "main"),
             path=data.get("path", "marketplace.json"),
+            source_type="github",
         )
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply src/apm_cli/marketplace/models.py:95]: Fixed. `from_dict()` now raises `ValueError` if `source_type="url"` with empty/missing url. Added 2 tests. See commit fa8b7a4.

File: src/apm_cli/marketplace/models.py [hunk 10]
```diff
@@ -226,4 +271,123 @@ def parse_marketplace_json(
...
+                source_name,
+            )
+            continue
+        description = entry.get("description", "")
+        plugins.append(
+            MarketplacePlugin(
+                name=name,
+                source={"type": skill_type, "url": url, "digest": digest},
+                description=description,
+                source_marketplace=source_name,
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply src/apm_cli/marketplace/models.py:406]: Fixed. Added validation: `type` must be a string in `("skill-md", "archive")`, `url` must be a non-empty string, `description` defaults to empty string if non-string. Invalid entries are skipped with a warning log. Added 7 tests. See commit fa8b7a4.

File: src/apm_cli/marketplace/archive.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        List of relative paths extracted.
+
+    Raises:
+        ArchiveError: On download failure, unrecognised format, or unsafe content.
+    """
+    try:
+        resp = requests.get(url, headers={"User-Agent": "apm-cli"}, timeout=60)
+        resp.raise_for_status()
+    except requests.exceptions.RequestException as exc:
+        raise ArchiveError(f"Failed to download archive from {url!r}: {exc}") from exc
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply src/apm_cli/marketplace/archive.py:232]: Fixed. Added HTTPS enforcement to `download_and_extract_archive()`: rejects non-HTTPS URLs up-front and validates post-redirect scheme. Added 3 tests including an HTTPS->HTTP redirect test. See commit fa8b7a4.

File: src/apm_cli/marketplace/archive.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+                real_dest = os.path.realpath(dest_dir)
+                if not dest_path.startswith(real_dest + os.sep) and dest_path != real_dest:
+                    raise ArchiveError(
+                        f"Archive member would extract outside destination: {member.name!r}"
+                    )
+
+                os.makedirs(os.path.dirname(dest_path), exist_ok=True)
+                with tf.extractfile(member) as src, open(dest_path, "wb") as dst:
+                    dst.write(src.read())
+                extracted.append(member.name)
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply src/apm_cli/marketplace/archive.py:154]: Fixed. Added `member.isreg()` check before extraction and null check on `extractfile()` return value. Non-regular members (device files, FIFOs) are now skipped with a warning. Added 4 tests. See commit fa8b7a4.

File: src/apm_cli/marketplace/archive.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+    normalized = os.path.normpath(member_path)
+    if normalized.startswith(".."):
+        raise ArchiveError(
+            f"Archive member path traversal detected: {member_path!r}"
+        )
+    parts = forward.split("/")
+    if ".." in parts:
+        raise ArchiveError(
+            f"Archive member path traversal detected: {member_path!r}"
+        )
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply src/apm_cli/marketplace/archive.py]: Fixed. Refactored `_check_archive_member()` to use `validate_path_segments()` from `path_security.py`. Refactored containment checks in `_extract_tar_gz()` and `_extract_zip()` to use `ensure_path_within()`. Archive-specific checks (null bytes, Windows UNC paths) remain as additional guards. See commit fa8b7a4.

File: tests/unit/marketplace/test_marketplace_archive.py [hunk 14]
```diff
@@ -0,0 +1,349 @@
+"""Tests for archive download and extraction (Step 7 — gap #14).
+
+Covers: _check_archive_member safety checks, _detect_archive_format, _extract_tar_gz,
+_extract_zip, and the download_and_extract_archive public API.
+"""
+
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply tests/unit/marketplace/test_marketplace_archive.py]: Fixed. All em-dashes replaced with ASCII `--` across the file. Also replaced Unicode arrows and section signs in other test files. See commits fa8b7a4 and 516370a.

File: tests/unit/marketplace/test_marketplace_url_client.py [hunk 15]
```diff
@@ -0,0 +1,854 @@
+"""Tests for URL-based marketplace client fetch path.
+
+Covers: _cache_key() URL sources, _fetch_url_direct(), fetch_marketplace()
+URL branch, format auto-detection, cache read/write, stale-while-revalidate.
+GitHub paths are not touched — regression tests confirm they still work.
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply tests/unit/marketplace/test_marketplace_url_client.py]: Fixed. All em-dashes and Unicode arrows replaced with ASCII equivalents. See commits fa8b7a4 and 516370a.

File: tests/unit/marketplace/test_marketplace_url_commands.py [hunk 16]
```diff
@@ -0,0 +1,521 @@
...
+    ])
+    def test_mixed_case_scheme_detected_as_url(self, runner, url):
+        """URL detection must be case-insensitive per RFC 3986 §3.1."""
+        from apm_cli.commands.marketplace import marketplace
+
+        result = runner.invoke(marketplace, ["add", url])
+        # Should NOT produce "Invalid format ... OWNER/REPO" —
+        # it should enter the URL path (may fail on http or succeed on https)
+        assert "OWNER/REPO" not in result.output
+
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply tests/unit/marketplace/test_marketplace_url_commands.py:116]: Fixed. All em-dashes, Unicode arrows, and section signs replaced with ASCII. See commits fa8b7a4 and 516370a.

File: tests/unit/marketplace/test_marketplace_url_models.py [hunk 17]
```diff
@@ -0,0 +1,549 @@
+"""Tests for URL-based marketplace source and Agent Skills index parser.
+
+Covers MarketplaceSource URL fields, serialization round-trips, and the
+parse_agent_skills_index() parser including schema enforcement, skill name
+validation, and source type handling.
+"""
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply tests/unit/marketplace/test_marketplace_url_models.py:6]: Fixed. All em-dashes replaced with ASCII `--`. See commit fa8b7a4.

File: tests/unit/marketplace/test_marketplace_url_resolver.py [hunk 18]
```diff
@@ -0,0 +1,196 @@
+"""Tests for URL-source resolver behaviour.
+
+Covers:
+- resolve_plugin_source() with skill-md and archive Agent Skills types
+- _resolve_url_source() with non-GitHub HTTPS URLs
+- resolve_marketplace_plugin() end-to-end with a URL marketplace source
+"""
```
[the remark this answers is not in the document]
Comment (Vicente-Pastor) [reply tests/unit/marketplace/test_marketplace_url_resolver.py:7]: Fixed. All em-dashes replaced with ASCII `--`. See commit fa8b7a4.

File: src/apm_cli/marketplace/models.py (hunk 1 above)
Comment (Vicente-Pastor) [reply src/apm_cli/marketplace/models.py]: Fixed. All em-dashes replaced with `--` throughout models.py (lines 15, 30, 31 and others). Also swept all other PR files for em-dashes and non-ASCII chars (arrows, section signs). See commits fa8b7a4 and 516370a.

File: src/apm_cli/marketplace/archive.py (hunk 2 above)
Comment (Vicente-Pastor) [reply src/apm_cli/marketplace/archive.py:35]: Fixed. Refactored archive.py to import and use `validate_path_segments()` and `ensure_path_within()` from `path_security.py`. Archive-specific guards (null bytes, Windows UNC/drive-letter paths) are kept as additional checks on top of the centralized ones. See commit fa8b7a4.

File: src/apm_cli/marketplace/client.py (hunk 3 above)
Comment (Vicente-Pastor) [reply src/apm_cli/marketplace/client.py:295]: Fixed. Added `_MAX_INDEX_BYTES = 10 * 1024 * 1024` (10 MB) constant. Checks both `Content-Length` header before reading and actual body size after reading. Added 3 tests. See commit fa8b7a4.

File: src/apm_cli/commands/marketplace.py (hunk 4 above)
Comment (Vicente-Pastor) [reply src/apm_cli/commands/marketplace.py]: Fixed. Replaced Unicode em-dash with ASCII `-` in the Rich table output. See commit fa8b7a4.

**Comment by Vicente-Pastor** [comment]
## Addressing review body items

@sergio-sisternes-epam Thanks for the thorough review! All items have been addressed:

### Inline comments (C01-C19)
All 19 inline review comments have been fixed and replied to individually. Summary:
- **Em-dashes**: All non-ASCII characters replaced with ASCII equivalents across all PR files (also caught Unicode arrows and section signs)
- **Security**: HTTPS enforcement + post-redirect validation in archive.py, centralized path_security.py integration, 10MB response size limit, tar member type checks, field validation in parser
- **Correctness**: Case-insensitive scheme validation (urlparse-based), empty URL rejection in from_dict(), CodeQL fixes

### Review body items
1. **CHANGELOG.md** -- Added 6 entries under `[Unreleased] > Added` with (#676, #691) references
2. **Documentation** -- Updated `docs/src/content/docs/guides/marketplaces.md` (URL registration section), `docs/src/content/docs/reference/cli-commands.md` (URL syntax/examples), and `packages/apm-guide/.apm/skills/apm-usage/commands.md`
3. **Deleted TestCacheKey tests** -- Restored all 3 tests (`test_github_default_unchanged`, `test_non_default_host_includes_host`, `test_different_hosts_different_keys`)
4. **Browse command** -- Verified it works for URL sources (uses `get_marketplace_by_name` -> `fetch_marketplace` which handles URL sources transparently)
5. **Merge conflicts with #677** -- Noted. Will rebase after #677 merges

### Test status
387 tests passing (up from 384 before review fixes). All new tests follow AAA pattern with descriptive names.

See commits fa8b7a4 and 516370a.

**Review by danielmeppiel — CHANGES_REQUESTED** [review]
Filing a formal review to record the maintainer position alongside the reframe comment in #676 (https://github.com/microsoft/apm/issues/676#issuecomment-4277028253).

This PR's plumbing is the foundation we want to build on — `MarketplaceSource`, ETag/conditional-refresh, archive safety, auth-first client, lockfile provenance, the test scope. The destination contract is the change: reframed #676 targets parity with [Anthropic's `marketplace.json` spec](https://code.claude.com/docs/en/discover-plugins) (4 source shapes) instead of the Cloudflare `.well-known/agent-skills/` index.

Phasing as described in #676:
- **Phase 1** — three missing source shapes (git URL, local path, remote `marketplace.json` URL)
- **Phase 2** — auth integration via `AuthResolver` (covers your #692)

Holding this PR in `Changes requested` while we converge on Phase 1 acceptance criteria. Once we agree on direction (drop a 👍 on the #676 comment), happy to work with you on a re-aimed PR that keeps the plumbing and swaps the destination schema. The Sergio review changes are also worth incorporating in that pass.

No urgency on a force-march here — let's land the right scope, not the fast one.

**Comment by danielmeppiel** [comment]
Closing this with thanks for the substantial foundation work -- your `MarketplaceSource` generalization, ETag/conditional-refresh design, archive safety, and lockfile provenance fields directly informed #1476, which merged today and now provides `apm marketplace add` for local paths, `file://` URIs, SSH URLs, and HTTPS URLs to any git host. That covers 3 of the 4 source shapes from #676.

The 4th case (remote `marketplace.json` URL) is not closed by #1476, but we don't want to build against the Agent Skills Discovery RFC that this PR targets: the org `agentskills/` on GitHub doesn't exist, `schemas.agentskills.io` doesn't resolve, and the agentskills.io spec itself only covers `SKILL.md` (no canonical discovery wire format). When this case is needed, the right shape to follow is Anthropic's `marketplace.json` schema (already partly modeled by #1476), not the v0.2.0 schema cited here.

Your authorship is credited in #1476. Thank you.
